src/utils/intrusive_ptr.h

```diff
@@ -0,0 +1,81 @@
+//-----------------------------------------------------------------------------
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//-----------------------------------------------------------------------------
+#ifndef SRC_UTILS_INTRUSIVE_PTR_H_
+#define SRC_UTILS_INTRUSIVE_PTR_H_
+
+#include "src/utils/ref_counted.h"
+
+namespace raksha {
+
+template <typename T>
+class intrusive_ptr {
```

Comment:
Not wild about the name `intrusive_ptr`; it's long, and it doesn't describe *why* we are intruding into the pointed-at memory. Can we make it something like `refct_ptr` instead?

Reply:
It is somewhat standard term used for objects with embedded reference counts: https://www.boost.org/doc/libs/1_63_0/libs/smart_ptr/intrusive_ptr.html

I will add a comment that this is inspired by boost (which is true).

Reply:
ok sounds good